File: Java/UnicodeTests.java
import java.text.BreakIterator;
import java.text.Collator;
import java.text.Normalizer;
import java.text.Normalizer.Form;
import java.util.Locale;

public class UnicodeTests {

  static int passed = 0;
  static int failed = 0;
  static String testName;
  
  static void test(String name, boolean result) {
    if (result) {
      passed++;
    } else {
      failed++;
      System.out.println("Test " + name + " failed");
    }
  }

  static int countGraphemes(String input) {
    BreakIterator graphemes = BreakIterator.getCharacterInstance();
    graphemes.setText(input);
    int n = 0;
    while (graphemes.next() !=  BreakIterator.DONE) {
        n++;
    }
    return n;
  }

  static String reverseGraphemes(String input) {
    BreakIterator graphemes = BreakIterator.getCharacterInstance();
    graphemes.setText(input);
    StringBuilder result = new StringBuilder();
    int end = graphemes.last();
    int beg = graphemes.previous();
    for (; beg != BreakIterator.DONE; end = beg, beg = graphemes.previous()) {
        result.append(input.substring(beg, end));
    }
    return result.toString();
  }

  static String substringGraphemes(String input, int beg, int end) {
    BreakIterator graphemes = BreakIterator.getCharacterInstance();
    graphemes.setText(input);
    return input.substring(graphemes.next(0), graphemes.next(end));
  }

  public static void main(String[] args) {
    
    String unicodeString;
    Locale turkish = new Locale("tr", "TR");
    Locale german = new Locale("de", "DE");
    
    test("Unicode 1", Normalizer.normalize("c\u0327", Normalizer.Form.NFC).equals("\u00E7"));
    
    test("Unicode 2", !"c\u0327".equals("\u00E7"));
    
    test("Unicode 3", countGraphemes("noe\u0308l") == 4);
    
    test("Unicode 4", reverseGraphemes("noe\u0308l").equals("le\u0308on"));
    
    test("Unicode 5", substringGraphemes("noe\u0308l", 0, 3).equals("noe\u0308")
                      && countGraphemes(substringGraphemes("noe\u0308l", 0, 3)) == 3);
            
    test("Unicode 6", "ba\uFB04e".toUpperCase().equals("BAFFLE"));
    
    test("Unicode 7", "cant\u00F9".toUpperCase().equals("CANT\u00D9"));
    
    test("Unicode 8", "cantu\u0300".toUpperCase().equals("CANTU\u0300"));
    
    unicodeString = "\uD834\uDD1E - The Treble Clef";
    unicodeString = unicodeString.replace("\uD834\uDD1E", "\uD834\uDD22");
    test("Unicode 9", unicodeString.replace("Treble", "Bass")
                                    .equals("\uD834\uDD22 - The Bass Clef"));
    
    test("Unicode 10", "\uD834\uDD22 - The Bass Clef".codePointCount(
                        0, "\uD834\uDD22 - The Bass Clef".length()) == 17 );
                                                      
    
    test("Unicode 11", "i".toUpperCase(turkish).equals("\u0130"));
    
    test("Unicode 12", "I".toLowerCase(turkish).equals("\u0131"));
    
    test("Unicode 13", "stra\u00DFe".toUpperCase().equals("STRASSE"));
    
    test("Unicode 14", countGraphemes(Normalizer.normalize("\u03C8\u3099", 
                                      Normalizer.Form.NFC)) == 1);
    
    test("Unicode 15", countGraphemes(
                          Normalizer.normalize("e\u0308\uD834\uDD1E\u03C8\u3099", 
                          Normalizer.Form.NFC)) == 3); 
    
    Collator deCollator = Collator.getInstance(german);
    test("Unicode 16", deCollator.compare("weiss", "weiß") == 0);
    
    System.out.println("Tests performed: " + (passed + failed));
    System.out.println("Tests passed:    " + passed);
    System.out.println("Tests failed:    " + failed);
  
  }
  
}

File: C#/UnicodeTests.cs
using System;
using System.Globalization;
using System.Text;
 
public class UnicodeTests {
    private static int failed = 0;
    private static int passed = 0;
    
    private static String nfc(String input) {
        return input.Normalize(NormalizationForm.FormC);
    }

    private static String nfd(String input) {
        return input.Normalize(NormalizationForm.FormD);
    }

    private static String nfkd(String input) {
        return input.Normalize(NormalizationForm.FormKD);
    }

    private static string reverseGraphemes(string uS) {
      string sU = "";
      TextElementEnumerator text = StringInfo.GetTextElementEnumerator(uS);
      text.Reset();
      while (text.MoveNext()) {
        sU = text.GetTextElement() + sU;
      }
      return sU;
    }
    
    private static string substringGraphemes(string uS, int beg, int end) {
    TextElementEnumerator text = StringInfo.GetTextElementEnumerator(uS);
    string subSU = "";
    text.Reset();
    for (int i = 0; i < beg; i++) {
      if (!text.MoveNext()) {return subSU;} 
    }
    for (int i = beg; i < end; i++) {
      if (!text.MoveNext()) {return subSU;}
      subSU = subSU + text.GetTextElement();
    }
    return subSU;
    }
    
    private static void test(string name, bool result) {
        if (result) {
            passed++;
        } else {
            failed++;
            Console.WriteLine("Test " + name + " failed");
        }
    }
    
    static public void Main () {
    
        string uS;
        StringInfo sI;
    
        test("Unicode 1", nfc("c\u0327") == "\u00E7");
        
        test("Unicode 2", "c\u0327" != "\u00E7");
        
        sI = new StringInfo("noe\u0308l");
        test("Unicode 3", sI.LengthInTextElements == 4);
        
        test("Unicode 4", reverseGraphemes("noe\u0308l") == "le\u0308on");
        
        string subUS = substringGraphemes("noe\u0308l", 0, 3);
        sI = new StringInfo(subUS);
        test("Unicode 5", (subUS == "noe\u0308") && (sI.LengthInTextElements == 3));
          
        test("Unicode 6", "ba\uFB04e".ToUpper() == "BAFFLE");
        
        test("Unicode 7", "cant\u00F9".ToUpper() == "CANT\u00D9");

        test("Unicode 8", "cantu\u0300".ToUpper() == "CANTU\u0300");
        
        uS = "\uD834\uDD1E - The Treble Clef";
        uS = uS.Replace("\uD834\uDD1E", "\uD834\uDD22");
        test("Unicode 9", uS.Replace("Treble", "Bass") == "\uD834\uDD22 - The Bass Clef");
        
        sI = new StringInfo("\uD834\uDD22 - The Bass Clef");
        test("Unicode 10", sI.LengthInTextElements == 17);
        
        test("Unicode 11", "i".ToUpper(new CultureInfo("tr-TR", false)) == "\u0130");
        
        test("Unicode 12", "I".ToLower(new CultureInfo("tr-TR", false)) == "\u0131");

        test("Unicode 13", "stra\u00DFe".ToUpper() == "STRASSE");
      
        sI = new StringInfo(nfc("\u03C8\u3099"));
        test("Unicode 14", sI.LengthInTextElements == 1);
        
        sI = new StringInfo(nfc("e\u0308\uD834\uDD1E\u03C8\u3099"));
        test("Unicode 15", sI.LengthInTextElements == 3);
        
        test("Unicode 16", String.Compare("weiss", "weiß") == 0);
        
        test("Unicode 17", nfd("e\u0303\u033D\u032A") == "e\u032A\u0303\u033D");

        test("Unicode 18", "XII" == nfkd("\u216B"));

        test("Unicode 19", "XII" != "\u216B");
    
        test("Unicode 20", "\u1E14" == nfc("\u1E14"));
    
        test("Unicode 21", "E\u0304\u0300" == nfd("\u1E14"));
    
        test("Unicode 22", "\u1E14" == nfc("\u0112\u0300"));
    
        test("Unicode 23", "E\u0304\u0300" == nfd("\u0112\u0300"));   
   
        test("Unicode 24", "\u00C8\u0304" == nfc("\u00C8\u0304"));
        
        test("Unicode 25", nfd("e\u0303\u033D\u032A")!= "e\u033D\u032A\u0303");
            
	      Console.WriteLine("Number of Tests   " + (passed + failed));
        Console.WriteLine("Passed            " + passed);
        Console.WriteLine("Failed            " + failed);
    }
}